Source/bmalloc/libpas/src/libpas/pas_stats.h

```diff
@@ -0,0 +1,311 @@
+ * couldn't find a way to do so ergonomically.
+ *
+ * Since this framework is intended for use inside of an allocator, it is
+ * intended to have low overhead (both for enabled and disabled counters) and to
+ * have minimal use of heap-allocated memory -- however, there is room for
+ * improvement on both counts.
+ *   - Re.: heap-allocated memory, on the logging path we currently do rely on
+ *     heap allocations to make it easier for people to add new counters, as
+ *     using a fixed-size static allocation per counter would mean every counter
+ *     would need to pre-compute the theoretical maximum size of its json
+ *     payload.
+ *     Normally the utility heap would be a good fit for this use-case, but to
+ *     avoid reentrancy we do not use libpas to allocate this memory -- even by
+ *     going through the system heap. Instead, we call `malloc` directly. These
+ *     buffers are cached so it shouldn't happen often but it would be better to
+ *     be able to remove that dependency.
+ *   - Re.: performance, the current design is not bad but does introduce a lot
+ *     of atomic traffic and cross-core contention. Ideally, we would instead
+ *     have a per-thread 'local stat counter cache' which we would then
+ *     periodically accumulate into a global stat-counter object. Individual
+ *     stat counters would have to be aware since they need to implement their
+ *     own accumulate functions.
+ *     Even better than thread-local would be if we had something like Linux'
+ *     rseqs, as we could then store this data per-CPU and avoid any migration
+ *     whatsoever.
+ *     In both cases though, we would risk under-counting statistics unless we
+ *     implemented an analog of what pas-TLCs do where they iterate over other
+ *     threads' TLCs and collect data out. Doing so generically across all kinds
+ *     of stat counters seems like a challenge.
+ */
+
+#include "pas_config.h"
+
+#include "pas_allocation_mode.h"
+#include "pas_lock.h"
+#include "pas_internal_config.h"
+#include "pas_utils.h"
+#include <stdbool.h>
+#include <stdatomic.h>
+#include <stdint.h>
+#include <stdio.h>
+
+PAS_BEGIN_EXTERN_C;
+
+/**********************************/
+/* General stat-counter machinery */
+/**********************************/
+
+typedef char* (*pas_stats_stat_logger)(void* stat);
+
+/* Statistics sink for output */
+typedef struct pas_stats_sink pas_stats_sink;
+typedef void (*pas_stats_sink_output_func)(pas_stats_sink* sink, const char* json_output);
+
+// buf and len are always owned by pas_stats_ensure_print_buffer, and the
+// underlying storage should be considered an implementation detail.
+typedef struct pas_stats_print_buffer {
+    char* ptr;
+    size_t len;
+} pas_stats_print_buffer;
+char* pas_stats_ensure_print_buffer(pas_stats_print_buffer*, size_t desired_len);
+
+// All stat-counter structs must have a field of this type named '.data' as
+// their first member
```

Comment:
Is this reference to '.data' stale?  I don't see any reference to a `data` field anywhere in the patch.

Reply:
Not stale, I just mistyped this for whatever reason